lib/base/networking/apis/tumdev/cached_client.dart

```diff
@@ -0,0 +1,159 @@
+
+  CachedCampusClient.mobileCache(Directory directory)
+      : super(_channel(),
+      options: _callOptions(),
+      interceptors: [CacheInterceptor.mobileCache(directory)]);
+
+  void invalidateCache() {
+    // TODO: implement interceptUnary
+    throw UnimplementedError();
+  }
+
+  static CallOptions _callOptions() {
+    return CallOptions(metadata: {
+      "x-app-version": "unknown",
+      "x-app-build": "unknown",
+      "x-device-id": "unkonwn",
+      "x-os-version": "unknown",
+    });
+  }
+
+  static GrpcOrGrpcWebClientChannel _channel() {
+    return GrpcOrGrpcWebClientChannel.toSeparateEndpoints(
+        grpcHost: "api.tum.app",
+        grpcPort: 443,
+        grpcTransportSecure: true,
+        grpcWebHost: "api-grpc.tum.app",
+        grpcWebPort: 443,
+        grpcWebTransportSecure: true);
+  }
+}
+
+class CacheInterceptor implements ClientInterceptor {
+  late HiveCacheStore hiveCacheStore;
+  late LRUMemoryCacheStore memCacheStore;
+
+  CacheInterceptor.mobileCache(Directory directory) {
+    hiveCacheStore = HiveCacheStore(directory.path);
+  }
+  CacheInterceptor.webCache() {
+    memCacheStore = LRUMemoryCacheStore();
+  }
+
+  @override
+  ResponseStream<R> interceptStreaming<Q, R>(
+      ClientMethod<Q, R> method,
+      Stream<Q> requests,
+      CallOptions options,
+      ClientStreamingInvoker<Q, R> invoker) {
+    final key = "${method.path}?${requests.toString()}";
+    final (bool, dynamic) cachedResponse;
+    if (kIsWeb) {
+      cachedResponse = hiveCacheStore.get(key);
+    } else {
+      cachedResponse = memCacheStore.get(key);
+    }
+    if (cachedResponse.$1 && cachedResponse.$2 != null) {
+      return ResponseStream(cachedResponse.$2);
+    }
+
+    final response = invoker(method, requests, options);
+    response.listen((data) {
+      if (kIsWeb) {
+        memCacheStore.put(key, data);
+      } else {
+        hiveCacheStore.put(key, data);
+      }
+    });
+
+    return response;
+  }
+
+  @override
+  ResponseFuture<R> interceptUnary<Q, R>(
+      ClientMethod<Q, R> method,
+      Q request,
+      CallOptions options,
+      ClientUnaryInvoker<Q, R> invoker,
+      ) {
+    final key = "${method.path}?${request.toString()}";
+    final (bool, dynamic) cachedResponse;
+    if (kIsWeb) {
+      cachedResponse = hiveCacheStore.get(key);
+    } else {
+      cachedResponse = memCacheStore.get(key);
+    }
+    if (cachedResponse.$1 && cachedResponse.$2 != null) {
+      return ResponseFuture(cachedResponse.$2);
+    }
+
+    // If not found in cache, invoke the actual RPC and cache the response
+    final response = invoker(method, request, options);
+    response.then((data) {
+      if (kIsWeb) {
+        memCacheStore.put(key, data);
+      } else {
+        hiveCacheStore.put(key, data);
+      }
+    });
+
+    return response;
+  }
+}
+
+class HiveCacheStore {
+  final String _cachePath;
+  final Duration ttl;
+  HiveCacheStore(this._cachePath, {this.ttl=const Duration(days: 30)});
```

Comment:
The behaviour should be the same as for the other APIs, if the app detects that the user is connected the HiveCacheStore checks if the cached response is older than 10 minutes, if that's not the case a refetch should happen.